### DISCUSSION TITLE
Release 0.21

### DISCUSSION THREAD

**Discussion by viktormalik** [bpftrace/bpftrace#3191] [Development]
We should do a new release as it's been a while since we did one and we have a lot of new content.

This discussion is to gather the list of open MRs/issues that need to be addressed before doing the release:
- #3149
- #3024 
- #3182
- #3164

Please add other issues that you think need addressing.

**Comment by ajor** [comment]
We'll need to do something about this one too:
- #3021

It means that for-loops can trigger verifier errors if used in multiple probes.

I know you wanted to wait one more release before we switch to libbpf-based program loading, so we'll have to come up with some alternative. In order of my preferences:
1. Implement some workaround as discussed in the issue. Sounds like there were problems with this approach.
2. Give a nicer error message to users when attempting to use for-loops in multiple probes.
3. Remove documentation for for-loops until this issue is resolved.

I'll investigate.

**Reply by viktormalik** [reply]
Good point. @jordalgo did spend some time trying to find a workaround but unfortunately wasn't successful. It seems to be quite hard to persuade libbpf to load an ELF with static functions in sections other than `.text`.

So, my preference would be option 2.

**Reply by viktormalik** [reply]
I implemented option 2 in #3233.

**Comment by jordalgo** [comment]
Didn't quite remove `LOG(FATAL)` yet - at least not until this gets merged 😉 https://github.com/bpftrace/bpftrace/pull/3164

**Comment by viktormalik** [comment]
@ajor what's the status of #3182? AFAIK, it's the last issue preventing us from doing a release.

**Reply by ajor** [reply]
Sorry, I was off last week - thanks for picking it up!

**Comment by viktormalik** [comment]
In order to move forward with the release, I implemented a nicer error message for #3182. So, the remaining PRs to be merged before the release can be done are:
- #3233
- #3246 
- #3250

**Reply by jordalgo** [reply]
+1 - Let's get these merged. We haven't done a release in a LONG time.

**Comment by ajor** [comment]
Idea: should we cut a release branch for 0.21 ahead of the actual release?

That way we'll have a "stable" base which we can backport the necessary patches to and we won't include cutting edge features in the release or block other changes from getting merged into master.

With the current approach of using a snapshot of master as the release once our blocking issues are fixed there, we're ending up including a lot of new features into 0.21 the longer we wait.  I think it'd be good to give some features a bit of time to be tested before making it into a release.

**Reply by viktormalik** [reply]
This is a great idea. I agree with the reasoning behind, we're getting too many new features into the release while waiting for blocking bug fixes.

On the other hand, it's not as simple as just cutting a new branch. We at least need to update the CHANGELOG so that the entries in the current `Unreleased` section are moved into the upcoming release section. Also tags for the releases done this way will not be seen when doing `git log` on master but I'm not sure if that causes any real trouble (I use the tags as anchor points when bisecting a bug but that's just out of convenience).
